### ISSUE TITLE
Design question: Should the MsSql CreateSchema() method really be async?

### ISSUE THREAD

**Issue by BennieCopeland** [damianh/SqlStreamStore#66]
Currently the method is async, but when will this be called outside of an application's composition root, which is generally synchronous? I'm using SimpleInjector and it does not take async lambda expressions, so I specifically have to call Wait() since I want to make sure the database is ready before I start calling into it.

```csharp
container.Register<IStreamStore>(() =>
{
    var connectionString = this.Configuration.GetConnectionString("EventStore");
    var settings = new MsSqlStreamStoreSettings(connectionString);

    var store = new MsSqlStreamStore(settings);
    store.CreateSchema().Wait();

    return store;
}, Lifestyle.Singleton);
```

**Comment by bartelink** [comment]
If you have assumptions like this in your system (initializing async stuff), you're best off surfacing them as an initialization phase with clear parallelization and error management.

Anything in a lib like this that can be async should be - search around for Sync over Async is bad for why.

One way of managing async initialization, and rendezvousing with same can be to [wrap the work in a `LazyTask`](http://stackoverflow.com/a/37390850/11635) - but even that does not beat you doing the init in a controlled fashion and then exposing a simple API.

As with [`LazyTask`](http://stackoverflow.com/questions/33872588/caching-the-result-from-a-n-async-factory-method-iff-it-doesnt-throw/33946166#33946166), for any Singleton lifestyle like this with significant initialization, in an actual app,  you want to know when and where this is happening, when it times out, when it has initialization problems, and who is queueing up waiting for it to complete.

So, in short, it should definitely be `async`, and that does _not_ imply `Wait()` (which is a deadlock `Wait`ing to happen).

**Comment by BennieCopeland** [comment]
@bartelink So if I'm understanding you correctly, I should move the creation of the store outside of the container itself into some precomposition initiliziation step, and then register the created store. That definitely makes sense for the reasons you provided. So something like:

``` csharp
IStreamStore store = InitializationWithErrorHandlig();

container.Register<IStreamStore>(store);
```

While I agree now about the reasons for it being async, I still don't see away around calling Wait() somewhere in the chain. The ```Startup.cs``` file in ASP.Net Core does not support async initialization at the moment. If I want to ensure that the store is configured prior to ASP.Net accepting web requests, I have to block startup until it finishes.

**Comment by bartelink** [comment]
@BennieCopeland Yes, somebody somewhere will need to wait in some manner - and you need to pick what your server is going to do if the init fails - but you did already have that problem - you're just confronting it explicitly.

One way to address it is by using [`CachedLazyTask.Fetch()`](http://stackoverflow.com/a/33872589/11635), stashing the task _but not `await` it_, as a way of kicking off the initialization.

Then in any flow that requires the DB to be initialized, `await` the task you stashed as a prelude to doing the stream-access stuff. IOW have a Store wrapper with a `Task Initialize()` which holds the init task - use that in your startup logic (if you feel inclined, you can even use your container's way of triggering an initialize to cause it to be triggered). Then consumers either get and `await` that, or you have a `Task<T>  CreateConnection()` which internally does that.

However, if you leave it that way, some random call (_or concurrent set of calls_) will be impacted by a random delay (and will see any `Exception` incurred by the initialization logic). Whether that makes sense depends on your context.

The other thing to consider is that you may end up with some form of gateway in any case which will deal with allowing connection string reconfig and/or connection caching etc., which this can be integrated into.

And, for everyone's sanity, all of this stuff should be explicit code outside your DI container, no matter how tempting [toy :trollface: ] examples from containers look.
----

All of this generalises to having
1) a warmup/dependency verification phase for your system (calling any static ctors which might end up in a bad state), which can be wired into app container warmup facilities
2) a `/status` endpoint which can re-verify.

**Comment by damianh** [comment]
Thanks for the response @bartelink , spot on :)

As what @bartelink says, I too won't recommend calling `store.CreateSchema()` every time your applications starts, and definitely don't do that inside a container. It's more a deployment concern. 

There is another thing to consider - `CreateSchema` will require  higher level of permissions (DDL) than the rest of the operations (DML) of the library. Very often one will want different connection string & credentials when modifying a db structure to the application that is read/writing data.

Personally I tend to do self-hosted app servers encapsulated in a single exe. Off this I will have various CLI commands. For example:

 1. `appserver.exe init-stream-store --connectionSting='...'` -> creates the schema using the supplied connection string, etc.
 2. `appserver.exe install` -> installs windows service.
 3. etc

> And, for everyone's sanity, all of this stuff should be explicit code outside your DI container,

+100

**Comment by BennieCopeland** [comment]
@damianh That sounds like a really good idea. I was wondering how to achieve something like that with a standard IIS site, but going down the service path would also alleviate my concerns of using background tasks to generate read models. I would love to see some sample code from both you and @bartelink to better understand your solutions. For the windows service, are you using TopShelf or installing the service yourself?

**Comment by bartelink** [comment]
@BennieCopeland My existing usage of the technique is closed source at a prev employer (but the linked SO posts show decent enough samples). My theories of what one might do in a web app type scenario are me trying to suggest indicative possibilities.

Damian's DDL permissions point is of paramount importance for anything real. While some test rig getting an in memory or other ephemeral store may indeed be assumed to give you sufficient access such that one can provision the schema from inside the app, this simply will not be the case in any actual app.

As you mention, you also [dont want to drive anything from background tasks in an app](https://www.hanselman.com/blog/HowToRunBackgroundTasksInASPNET.aspx) anyway. Basically the core bit to take from my ramblings is to have a status page and let it assert configuration is OK (and perhaps make server startup contingent on the same server env is happy checks).

Actually provisioning for a real app is part of the initializing the environment bit

- for an acceptance test, spin up some in memory DB (in a Class/Collection Fixture async setup step*) / attach to a SQL Server LocalDb template
- for some containerized or cloud scenario, supply a pointer to the store in question
- for a desktop app, the installer will likely set up the DB server and create the DB as an admin

In xUnit 2, you might think a Collection Fixture is the way to go, as it would provide a neat place to initialize a store, but tests needing the same Collection Fixture will never run in parallel. Hence you might instead opt to create a Class Fixture which uses something like `LazyTask` to coordinate exactly once initialization.
